File: oio337a/Algorithm_1/day_12/198.py
from typing import List
import copy


class Solution:
    def rob(self, nums: List[int]) -> int:
        '''
        현재를 선택하거나, 저번 것을 계속 가거나
        '''
        n = len(nums)
        if not nums:
            return 0
        dp = copy.deepcopy(nums)
        for i in range(1, n):
            if i == 1:
                dp[1] = max(dp[0], dp[1])
            else:
                dp[i] = max(dp[i - 1], dp[i] + dp[i - 2])
        return dp[n - 1]


nums = [1, 2, 3, 3, 1, 6]
test = Solution()
print(test.rob(nums))

File: saewoo1/day12/198.java
class Solution {
    public int rob(int[] nums) {
        int maxMoney = 0;
        if (nums.length == 1)
            return nums[0];
        if (nums.length == 2)
            return Math.max(nums[0], nums[1]);
        maxMoney = getMoney(nums);
        return maxMoney;
    }
    private int getMoney(int[] nums) {
        int[] dp = new int[nums.length];
        int max = 0;
        dp[0] = nums[0];
        dp[1] = nums[1];
        dp[2] = nums[2] + nums[0];
        max = Math.max(dp[1], dp[2]);
        for (int i = 3; i < dp.length; i++) {
            // dp[i] = Math.max(nums[i] + dp[i - 2], nums[i] + dp[i - 3]);
            dp[i] = Math.max(dp[i - 2], dp[i - 3]) + nums[i];
            max = Math.max(max, dp[i]);
        }
        return max;
    }
}